R/CsvHelper.R

```diff
@@ -0,0 +1,125 @@
+#' for uploading to a results database.
+#' 
+#' @param x  A data frame or tibble to write to disk.
+#' 
+#' @param file  The csv file to write.
+#' 
+#' @param warnOnUploadRuleViolations When TRUE, this function will provide
+#' warning messages that may indicate if the data is stored in a format in the
+#' CSV that may cause problems when uploading to a database.
+#' 
+#' @return 
+#'  Returns the input x invisibly.
+#'
+#' @export
+writeCsv <- function(x, file, warnOnUploadRuleViolations = TRUE) {
+  if (warnOnUploadRuleViolations) {
+    # Check the data.frame to ensure it is in the proper format for upload
+    isFormattedForDatabaseUpload(x)
+    
+    # Check if the file name is in lower snake case format
+    fileName <- gsub(x = basename(file), pattern = ".csv", replacement = "")
+    if (!isSnakeCase(fileName)) {
+      warning(paste("The filename:", basename(file), "is not in snake case format"))
+    }
+    
+    # Also perform a check to see if the fileName end in "s" which might indicate
+    # that the resulting file name is plural
+    if (endsWith(fileName, "s")) {
+      message(paste("The filename:", basename(file), "may be plural since it ends in 's'. Please ensure you are using singular nouns for your file names."))
+    }
+  }
+  
+  # Write the file
+  readr::write_csv(x = x,
+                   file = file)
+  invisible(x)
+}
+
+
+#' Is the data.frame formatted for uploading to a database?
+#'
+#' @description
+#' This function is used to check a data.frame to ensure all 
+#' column names are in snake case format. 
+#' 
+#' @param x  A data frame
+#' @param warn When TRUE, display a warning of any columns are not in snake 
+#' case format
+#' 
+#' @return 
+#' Returns TRUE if all columns are snake case format. If warn == TRUE,
+#' the function will emit a warning on the column names that are not in snake 
+#' case format.
+#'
+#' @export
+isFormattedForDatabaseUpload <- function(x, warn = TRUE) {
+  checkmate::assert_data_frame(x)
+  columnNames <- colnames(x)
+  columnNamesInSnakeCaseFormat <- isSnakeCase(columnNames)
+  if (!all(columnNamesInSnakeCaseFormat) && warn) {
+    problemColumns <- columnNames[!columnNamesInSnakeCaseFormat]
+    problemColumnsWarning <- paste(problemColumns, collapse = ", ")
+    problemColumnsFixedWarning <- paste(snakecase::to_snake_case(problemColumns), collapse = ", ")
+    warning(paste("The following data.frame column names are not in snake case format:", problemColumnsWarning, "\n  Consider revising the column names to:", problemColumnsFixedWarning))
+  }
+  return(all(columnNamesInSnakeCaseFormat))
+}
+
+#' Used to check if a string is in snake case
+#'
+#' @description
+#' This function is used check file and field names
+#' to ensure the proper casing is in use.
+#' 
+#' @param x  The string to evaluate
+#' 
+#' @return 
+#'  TRUE if the string is in snake case
+#'
+#' @export
+isSnakeCase <- function(x) {
+  y <- snakecase::to_snake_case(x, numerals = "asis")
```

Comment:
I would add an entire package dependency just to check if a string is snake case. A simple regex could do that, right?

Reply:
Yes indeed. I'm also using the `snakecase` package to suggest how the column names could be adjusted to use snakecase. That package offered some additional benefits in terms of automatically preserving names that were already in snakecase, etc. I'm not married to using it but thought it was potentially useful.

Reply:
I've removed the dependency on snakecase and opted for the regex approach.